src/core/qgsabstractdatabaseproviderconnection.cpp

```diff
@@ -0,0 +1,280 @@
+
+ ---------------------
+ begin                : 2.8.2019
+ copyright            : (C) 2019 by Alessandro Pasotti
+ email                : elpaso at itopen dot it
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+#include "qgsabstractdatabaseproviderconnection.h"
+#include "qgsexception.h"
+#include <QVariant>
+#include <QObject>
+
+QgsAbstractDatabaseProviderConnection::QgsAbstractDatabaseProviderConnection( const QString &name ):
+  QgsAbstractProviderConnection( name )
+{
+
+}
+
+QgsAbstractDatabaseProviderConnection::QgsAbstractDatabaseProviderConnection( const QString &name, const QString &uri ):
+  QgsAbstractProviderConnection( name, uri )
+{
+
+}
+QgsAbstractDatabaseProviderConnection::Capabilities QgsAbstractDatabaseProviderConnection::capabilities() const
+{
+  return mCapabilities;
+}
+
+void QgsAbstractDatabaseProviderConnection::checkCapability( QgsAbstractDatabaseProviderConnection::Capability capability ) const
+{
+  if ( ! mCapabilities & capability )
+  {
+    static QMetaEnum metaEnum = QMetaEnum::fromType<QgsAbstractDatabaseProviderConnection::Capability>();
+    const QString capName { metaEnum.valueToKey( capability ) };
+    throw QgsProviderConnectionException( QObject::tr( "Operation '%1' is not supported for this connection" ).arg( capName ) );
+  }
+}
+
+void QgsAbstractDatabaseProviderConnection::createVectorTable( const QString &schema,
+    const QString &name,
+    const QgsFields &fields,
+    QgsWkbTypes::Type wkbType,
+    const QgsCoordinateReferenceSystem &srs,
+    bool overwrite,
+    const QMap<QString, QVariant> *
+    options ) const
+{
+  Q_UNUSED( schema );
+  Q_UNUSED( name );
+  Q_UNUSED( fields );
+  Q_UNUSED( srs );
+  Q_UNUSED( overwrite );
+  Q_UNUSED( options );
+  Q_UNUSED( wkbType );
+  throw QgsProviderConnectionException( QObject::tr( "Operation 'createVectorTable' is not supported" ) );
+}
+
+void QgsAbstractDatabaseProviderConnection::renameTable( const QString &, const QString &, const QString & ) const
+{
+  checkCapability( Capability::RenameTable );
+}
+
+void QgsAbstractDatabaseProviderConnection::dropVectorTable( const QString &, const QString & ) const
+{
+  checkCapability( Capability::DropVectorTable );
+}
+
+bool QgsAbstractDatabaseProviderConnection::tableExists( const QString &schema, const QString &name ) const
+{
+  checkCapability( Capability::TableExists );
+  const auto constTables { tables( schema ) };
```

Comment:
auto -> ???

(another danger of using auto - makes PR reviewing very difficult!)

Reply:
come on ... 
```
$ grep -rl 'const auto const' src|wc -l
405
```

Reply:
Those will almost completely be a result of @m-kuhn's Q_FOREACH replacement script, or for iterators.

I'm still strongly of the feeling that auto should NEVER be used without exceptional reason, and doing so leads to fragile, unreadable code and ultimately regressions.